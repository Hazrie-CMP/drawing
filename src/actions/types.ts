import React from "react";
import { ExcalidrawElement } from "../element/types";
import { AppState } from "../types";

type ActionResult = {
  elements?: ExcalidrawElement[];
  appState?: AppState;
};

type ActionFn = (
  elements: readonly ExcalidrawElement[],
  appState: AppState,
  formData: any
) => ActionResult;

type UpdaterFn = (formData: any) => void;

export interface Action {
  name: string;
  PanelComponent?: React.FC<{
    elements: readonly ExcalidrawElement[];
    appState: AppState;
    updater: UpdaterFn;
  }>;
  perform: ActionFn;
  keyTest?: (event: KeyboardEvent) => boolean;
  contextItemLabel?: string;
}

export interface ActionsManagerInterface {
  actions: {
    [keyProp: string]: Action;
  };
  registerAction: (action: Action) => void;
  handleKeyDown: (
    event: KeyboardEvent,
    elements: readonly ExcalidrawElement[],
    appState: AppState
  ) => ActionResult | {};
}
