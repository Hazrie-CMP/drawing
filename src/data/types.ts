import { ExcalidrawElement } from "../element/types";
import { AppState, BinaryFiles, LibraryItems, LibraryItems_v1 } from "../types";
import type { cleanAppStateForExport } from "../appState";

export interface ExportedDataState {
  type: string;
  version: number;
  source: string;
  elements: readonly ExcalidrawElement[];
  appState: ReturnType<typeof cleanAppStateForExport>;
  files: BinaryFiles | undefined;
}

export interface ImportedDataState {
  type?: string;
  version?: number;
  source?: string;
  elements?: readonly ExcalidrawElement[] | null;
  appState?: Readonly<Partial<AppState>> | null;
  scrollToContent?: boolean;
  libraryItems?: LibraryItems | LibraryItems_v1;
  files?: BinaryFiles;
}

export interface ExportedLibraryData {
  type: string;
  version: number;
  source: string;
  libraryItems: LibraryItems;
  library?: LibraryItems | LibraryItems_v1;
}

export interface ImportedLibraryData extends Partial<ExportedLibraryData> {}
