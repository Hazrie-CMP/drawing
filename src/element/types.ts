import { newElement } from "./newElement";

export type ExcalidrawElement = ReturnType<typeof newElement>;
export type ExcalidrawTextElement = ExcalidrawElement & {
  type: "text";
  font: string;
  text: string;
  // for backward compatibility
  actualBoundingBoxAscent?: number;
  baseline: number;
};

export type ExcalidrawArrowElement = ExcalidrawElement & {
  type: "arrow";
  angle: number;
};

export type ExcalidrawLineElement = ExcalidrawElement & {
  type: "line";
  angle: number;
};
