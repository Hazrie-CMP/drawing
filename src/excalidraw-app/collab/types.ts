export enum UserIdleState {
  Active = "active",
  Idle = "idle",
  Away = "away",
}
