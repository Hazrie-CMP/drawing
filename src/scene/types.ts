import { ExcalidrawTextElement } from "../element/types";
import { AppState } from "../types";

export type SceneState = {
  scrollX: AppState["scrollX"];
  scrollY: AppState["scrollY"];
  // null indicates transparent bg
  viewBackgroundColor: string | null;
  zoom: number;
};

export type SceneScroll = {
  scrollX: AppState["scrollX"];
  scrollY: AppState["scrollY"];
};

export interface Scene {
  elements: ExcalidrawTextElement[];
}

export type ExportType = "png" | "clipboard" | "backend" | "svg";
