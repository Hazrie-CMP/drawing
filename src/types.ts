import { ExcalidrawElement } from "./element/types";

export type AppState = {
  draggingElement: ExcalidrawElement | null;
  resizingElement: ExcalidrawElement | null;
  // element being edited (e.g. text element)
  editingElement: ExcalidrawElement | null;
  elementType: string;
  exportBackground: boolean;
  currentItemStrokeColor: string;
  currentItemBackgroundColor: string;
  currentItemFont: string;
  viewBackgroundColor: string;
  scrollX: number;
  scrollY: number;
  cursorX: number;
  cursorY: number;
  name: string;
};
