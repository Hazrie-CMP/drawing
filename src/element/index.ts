import { ExcalidrawElement, Versioned, NonDeleted } from "./types";

export {
  newElement,
  newTextElement,
  newLinearElement,
  duplicateElement,
} from "./newElement";
export {
  getElementAbsoluteCoords,
  getCommonBounds,
  getDiamondPoints,
  getArrowPoints,
  getLinearElementAbsoluteBounds,
} from "./bounds";

export { handlerRectangles } from "./handlerRectangles";
export { hitTest } from "./collision";
export {
  resizeTest,
  getCursorForResizingElement,
  normalizeResizeHandle,
} from "./resizeTest";
export { isTextElement, isExcalidrawElement } from "./typeChecks";
export { textWysiwyg } from "./textWysiwyg";
export { redrawTextBoundingBox } from "./textElement";
export {
  getPerfectElementSize,
  isInvisiblySmallElement,
  resizePerfectLineForNWHandler,
  normalizeDimensions,
} from "./sizeHelpers";
export { showSelectedShapeActions } from "./showSelectedShapeActions";

export function getElementMap<TElement extends ExcalidrawElement>(
  elements: readonly TElement[],
) {
  return elements.reduce(
    (acc: { [key: string]: TElement }, element: TElement) => {
      acc[element.id] = element;
      return acc;
    },
    {},
  );
}

export function getDrawingVersion(
  elements: readonly Versioned<ExcalidrawElement>[],
) {
  return elements.reduce((acc, el) => acc + el.version, 0);
}

export function versionedToNonDeleted(
  elements: readonly Versioned<ExcalidrawElement>[],
): readonly NonDeleted<ExcalidrawElement>[] {
  return elements.filter(element => !element.isDeleted) as NonDeleted<
    ExcalidrawElement
  >[];
}

export function isNonDeleted(
  element: ExcalidrawElement,
): element is NonDeleted<ExcalidrawElement> {
  return !(element as Versioned<ExcalidrawElement>).isDeleted;
}

export function assertNonDeleted<TElement extends ExcalidrawElement>(
  element: TElement,
): asserts element is NonDeleted<TElement> {
  if (!isNonDeleted(element)) {
    throw new Error("Expected element not to be deleted");
  }
}
