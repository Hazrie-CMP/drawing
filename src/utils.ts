import {
  exportToCanvas as _exportToCanvas,
  exportToSvg as _exportToSvg,
} from "./scene/export";
import { getDefaultAppState } from "./appState";
import { AppState } from "./types";
import { getZoomOrigin } from "./scene";
import {
  CURSOR_TYPE,
  FONT_FAMILY,
  WINDOWS_EMOJI_FALLBACK_FONT,
} from "./constants";
import { ExcalidrawElement, FontFamily, FontString } from "./element/types";
import { getNonDeletedElements } from "./element";

export const SVG_NS = "http://www.w3.org/2000/svg";

let mockDateTime: string | null = null;

export const setDateTimeForTests = (dateTime: string) => {
  mockDateTime = dateTime;
};

export const getDateTime = () => {
  if (mockDateTime) {
    return mockDateTime;
  }

  const date = new Date();
  const year = date.getFullYear();
  const month = `${date.getMonth() + 1}`.padStart(2, "0");
  const day = `${date.getDate()}`.padStart(2, "0");
  const hr = `${date.getHours()}`.padStart(2, "0");
  const min = `${date.getMinutes()}`.padStart(2, "0");

  return `${year}-${month}-${day}-${hr}${min}`;
};

export const capitalizeString = (str: string) =>
  str.charAt(0).toUpperCase() + str.slice(1);

export const isToolIcon = (
  target: Element | EventTarget | null,
): target is HTMLElement =>
  target instanceof HTMLElement && target.className.includes("ToolIcon");

export const isInputLike = (
  target: Element | EventTarget | null,
): target is
  | HTMLInputElement
  | HTMLTextAreaElement
  | HTMLSelectElement
  | HTMLBRElement
  | HTMLDivElement =>
  (target instanceof HTMLElement && target.dataset.type === "wysiwyg") ||
  target instanceof HTMLBRElement || // newline in wysiwyg
  target instanceof HTMLInputElement ||
  target instanceof HTMLTextAreaElement ||
  target instanceof HTMLSelectElement;

export const isWritableElement = (
  target: Element | EventTarget | null,
): target is
  | HTMLInputElement
  | HTMLTextAreaElement
  | HTMLBRElement
  | HTMLDivElement =>
  (target instanceof HTMLElement && target.dataset.type === "wysiwyg") ||
  target instanceof HTMLBRElement || // newline in wysiwyg
  target instanceof HTMLTextAreaElement ||
  (target instanceof HTMLInputElement &&
    (target.type === "text" || target.type === "number"));

export const getFontFamilyString = ({
  fontFamily,
}: {
  fontFamily: FontFamily;
}) => {
  return `${FONT_FAMILY[fontFamily]}, ${WINDOWS_EMOJI_FALLBACK_FONT}`;
};

/** returns fontSize+fontFamily string for assignment to DOM elements */
export const getFontString = ({
  fontSize,
  fontFamily,
}: {
  fontSize: number;
  fontFamily: FontFamily;
}) => {
  return `${fontSize}px ${getFontFamilyString({ fontFamily })}` as FontString;
};

// https://github.com/grassator/canvas-text-editor/blob/master/lib/FontMetrics.js
export const measureText = (text: string, font: FontString) => {
  const line = document.createElement("div");
  const body = document.body;
  line.style.position = "absolute";
  line.style.whiteSpace = "pre";
  line.style.font = font;
  body.appendChild(line);
  line.innerText = text
    .split("\n")
    // replace empty lines with single space because leading/trailing empty
    //  lines would be stripped from computation
    .map((x) => x || " ")
    .join("\n");
  const width = line.offsetWidth;
  const height = line.offsetHeight;
  // Now creating 1px sized item that will be aligned to baseline
  // to calculate baseline shift
  const span = document.createElement("span");
  span.style.display = "inline-block";
  span.style.overflow = "hidden";
  span.style.width = "1px";
  span.style.height = "1px";
  line.appendChild(span);
  // Baseline is important for positioning text on canvas
  const baseline = span.offsetTop + span.offsetHeight;
  document.body.removeChild(line);

  return { width, height, baseline };
};

export const debounce = <T extends any[]>(
  fn: (...args: T) => void,
  timeout: number,
) => {
  let handle = 0;
  let lastArgs: T;
  const ret = (...args: T) => {
    lastArgs = args;
    clearTimeout(handle);
    handle = window.setTimeout(() => fn(...args), timeout);
  };
  ret.flush = () => {
    clearTimeout(handle);
    fn(...lastArgs);
  };
  return ret;
};

export const selectNode = (node: Element) => {
  const selection = window.getSelection();
  if (selection) {
    const range = document.createRange();
    range.selectNodeContents(node);
    selection.removeAllRanges();
    selection.addRange(range);
  }
};

export const removeSelection = () => {
  const selection = window.getSelection();
  if (selection) {
    selection.removeAllRanges();
  }
};

export const distance = (x: number, y: number) => Math.abs(x - y);

export const resetCursor = () => {
  document.documentElement.style.cursor = "";
};

export const setCursorForShape = (shape: string) => {
  if (shape === "selection") {
    resetCursor();
  } else {
    document.documentElement.style.cursor = CURSOR_TYPE.CROSSHAIR;
  }
};

export const isFullScreen = () =>
  document.fullscreenElement?.nodeName === "HTML";

export const allowFullScreen = () =>
  document.documentElement.requestFullscreen();

export const exitFullScreen = () => document.exitFullscreen();

export const getShortcutKey = (shortcut: string): string => {
  const isMac = /Mac|iPod|iPhone|iPad/.test(window.navigator.platform);
  if (isMac) {
    return `${shortcut
      .replace(/\bCtrlOrCmd\b/i, "Cmd")
      .replace(/\bAlt\b/i, "Option")
      .replace(/\bDel\b/i, "Delete")
      .replace(/\b(Enter|Return)\b/i, "Enter")}`;
  }
  return `${shortcut.replace(/\bCtrlOrCmd\b/i, "Ctrl")}`;
};
export const viewportCoordsToSceneCoords = (
  { clientX, clientY }: { clientX: number; clientY: number },
  appState: AppState,
  canvas: HTMLCanvasElement | null,
  scale: number,
) => {
  const zoomOrigin = getZoomOrigin(canvas, scale);
  const clientXWithZoom =
    zoomOrigin.x +
    (clientX - zoomOrigin.x - appState.offsetLeft) / appState.zoom;
  const clientYWithZoom =
    zoomOrigin.y +
    (clientY - zoomOrigin.y - appState.offsetTop) / appState.zoom;

  const x = clientXWithZoom - appState.scrollX;
  const y = clientYWithZoom - appState.scrollY;

  return { x, y };
};

export const sceneCoordsToViewportCoords = (
  { sceneX, sceneY }: { sceneX: number; sceneY: number },
  appState: AppState,
  canvas: HTMLCanvasElement | null,
  scale: number,
) => {
  const zoomOrigin = getZoomOrigin(canvas, scale);
  const x =
    zoomOrigin.x -
    (zoomOrigin.x - sceneX - appState.scrollX - appState.offsetLeft) *
      appState.zoom;
  const y =
    zoomOrigin.y -
    (zoomOrigin.y - sceneY - appState.scrollY - appState.offsetTop) *
      appState.zoom;

  return { x, y };
};

export const getGlobalCSSVariable = (name: string) =>
  getComputedStyle(document.documentElement).getPropertyValue(`--${name}`);

const RS_LTR_CHARS =
  "A-Za-z\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u02B8\u0300-\u0590\u0800-\u1FFF" +
  "\u2C00-\uFB1C\uFDFE-\uFE6F\uFEFD-\uFFFF";
const RS_RTL_CHARS = "\u0591-\u07FF\uFB1D-\uFDFD\uFE70-\uFEFC";
const RE_RTL_CHECK = new RegExp(`^[^${RS_LTR_CHARS}]*[${RS_RTL_CHARS}]`);
/**
 * Checks whether first directional character is RTL. Meaning whether it starts
 *  with RTL characters, or indeterminate (numbers etc.) characters followed by
 *  RTL.
 * See https://github.com/excalidraw/excalidraw/pull/1722#discussion_r436340171
 */
export const isRTL = (text: string) => {
  return RE_RTL_CHECK.test(text);
};

export function tupleToCoors(
  xyTuple: readonly [number, number],
): { x: number; y: number } {
  const [x, y] = xyTuple;
  return { x, y };
}

/** use as a rejectionHandler to mute filesystem Abort errors */
export const muteFSAbortError = (error?: Error) => {
  if (error?.name === "AbortError") {
    return;
  }
  throw error;
};

export const findIndex = <T>(
  array: readonly T[],
  cb: (element: T, index: number, array: readonly T[]) => boolean,
  fromIndex: number = 0,
) => {
  if (fromIndex < 0) {
    fromIndex = array.length + fromIndex;
  }
  fromIndex = Math.min(array.length, Math.max(fromIndex, 0));
  let i = fromIndex - 1;
  while (++i < array.length) {
    if (cb(array[i], i, array)) {
      return i;
    }
  }
  return -1;
};

export const findLastIndex = <T>(
  array: readonly T[],
  cb: (element: T, index: number, array: readonly T[]) => boolean,
  fromIndex: number = array.length - 1,
) => {
  if (fromIndex < 0) {
    fromIndex = array.length + fromIndex;
  }
  fromIndex = Math.min(array.length - 1, Math.max(fromIndex, 0));
  let i = fromIndex + 1;
  while (--i > -1) {
    if (cb(array[i], i, array)) {
      return i;
    }
  }
  return -1;
};

// Export functions

type ExportOpts = {
  elements: readonly ExcalidrawElement[];
  appState?: Omit<AppState, "offsetTop" | "offsetLeft">;
  getDimensions: (
    width: number,
    height: number,
  ) => { width: number; height: number; scale: number };
};

const exportToCanvas = ({
  elements,
  appState = getDefaultAppState(),
  getDimensions = (width, height) => ({ width, height, scale: 1 }),
}: ExportOpts) => {
  return _exportToCanvas(
    getNonDeletedElements(elements),
    { ...appState, offsetTop: 0, offsetLeft: 0 },
    {
      exportBackground: appState.exportBackground ?? true,
      viewBackgroundColor: appState.viewBackgroundColor ?? "#FFF",
      shouldAddWatermark: appState.shouldAddWatermark ?? false,
    },
    (width, height) => {
      const canvas = document.createElement("canvas");
      const ret = getDimensions(width, height);

      canvas.width = ret.width;
      canvas.height = ret.height;

      return canvas;
    },
  );
};

export const exportToBlob = (
  opts: ExportOpts & {
    mimeType?: string;
    quality?: number;
  },
): Promise<Blob | null> => {
  const canvas = exportToCanvas(opts);

  let { mimeType = "image/png", quality } = opts;

  if (mimeType === "image/png" && typeof quality === "number") {
    console.warn(`"quality" will be ignored for "image/png" mimeType`);
  }

  if (mimeType === "image/jpg") {
    mimeType = "image/jpeg";
  }

  quality = quality ? quality : /image\/jpe?g/.test(mimeType) ? 0.92 : 0.8;

  return new Promise((resolve) => {
    canvas.toBlob(
      (blob) => {
        resolve(blob);
      },
      mimeType,
      quality,
    );
  });
};

export const exportToSvg = ({
  elements,
  appState = getDefaultAppState(),
  exportPadding,
  metadata,
}: ExportOpts & {
  exportPadding?: number;
  metadata?: string;
}): SVGSVGElement => {
  return _exportToSvg(getNonDeletedElements(elements), {
    ...appState,
    exportPadding,
    metadata,
  });
};
