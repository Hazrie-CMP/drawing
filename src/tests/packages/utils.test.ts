import { ExcalidrawElement } from "../../element/types";
import * as utils from "../../packages/utils";

const elementFactory = (overrides = {}): ExcalidrawElement => ({
  id: "vWrqOAfkind2qcm7LDAGZ",
  type: "rectangle",
  x: 414,
  y: 237,
  width: 214,
  height: 214,
  angle: 0,
  strokeColor: "#000000",
  backgroundColor: "#15aabf",
  fillStyle: "hachure",
  strokeWidth: 1,
  strokeStyle: "solid",
  roughness: 1,
  opacity: 100,
  groupIds: [],
  strokeSharpness: "sharp",
  seed: 1041657908,
  version: 120,
  versionNonce: 1188004276,
  isDeleted: false,
  boundElementIds: null,
  ...overrides,
});
const diagramFactory = ({ overrides = {}, elementOverrides = {} } = {}) => ({
  type: "excalidraw",
  version: 2,
  source: "https://excalidraw.com",
  elements: [
    elementFactory({ type: "rectangle", ...elementOverrides }),
    elementFactory({ type: "diamond", ...elementOverrides }),
    elementFactory({ type: "ellipse", ...elementOverrides }),
  ],
  appState: {
    viewBackgroundColor: "#ffffff",
    gridSize: null,
  },
  ...overrides,
});

describe("exportToCanvas", () => {
  const EXPORT_PADDING = 10;
  it("with default arguments", () => {
    const canvas = utils.exportToCanvas({
      ...diagramFactory({ elementOverrides: { width: 100, height: 100 } }),
    });

    expect(canvas.width).toBe(100 + 2 * EXPORT_PADDING);
    expect(canvas.height).toBe(100 + 2 * EXPORT_PADDING);
  });

  it("when custom width and height", () => {
    const canvas = utils.exportToCanvas({
      ...diagramFactory({ elementOverrides: { width: 100, height: 100 } }),
      getDimensions: () => ({ width: 200, height: 200, scale: 1 }),
    });

    expect(canvas.width).toBe(200);
    expect(canvas.height).toBe(200);
  });
});

describe("exportToBlob", () => {
  describe("mime type", () => {
    afterEach(jest.restoreAllMocks);

    it("should change image/jpg to image/jpeg", async () => {
      const blob = await utils.exportToBlob({
        ...diagramFactory(),
        getDimensions: (width, height) => ({ width, height, scale: 1 }),
        mimeType: "image/jpg",
      });
      expect(blob?.type).toBe("image/jpeg");
    });

    it("should default to image/png", async () => {
      const blob = await utils.exportToBlob({
        ...diagramFactory(),
      });
      expect(blob?.type).toBe("image/png");
    });

    it("should warn when using quality with image/png", async () => {
      const consoleSpy = jest
        .spyOn(console, "warn")
        .mockImplementationOnce(() => void 0);

      await utils.exportToBlob({
        ...diagramFactory(),
        mimeType: "image/png",
        quality: 1,
      });

      expect(consoleSpy).toHaveBeenCalledWith(
        '"quality" will be ignored for "image/png" mimeType',
      );
    });
  });
});

describe("exportToSvg", () => {
  it("with default arguments", () => {
    const svgElement = utils.exportToSvg({
      ...diagramFactory({
        overrides: { appState: void 0 },
        elementOverrides: { width: 100, height: 100 },
      }),
    });

    expect(svgElement).toHaveAttribute("height", "120");
    expect(svgElement).toHaveAttribute("width", "120");
    expect(svgElement).toHaveAttribute("viewBox", "0 0 120 120");
    expect(svgElement).toMatchSnapshot();
  });
  it("with exportPadding and metadata", () => {
    const svgElement = utils.exportToSvg({
      ...diagramFactory({ elementOverrides: { width: 100, height: 100 } }),
      exportPadding: 0,
      metadata: "some metadata",
    });

    expect(svgElement.innerHTML).toMatch(/some metadata/);
    expect(svgElement).toHaveAttribute("height", "100");
    expect(svgElement).toHaveAttribute("width", "100");
    expect(svgElement).toHaveAttribute("viewBox", "0 0 100 100");
  });
});
