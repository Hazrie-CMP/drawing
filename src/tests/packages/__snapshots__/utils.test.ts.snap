// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exportToSvg with default arguments 1`] = `
Object {
  "collaborators": Map {},
  "currentChartType": "bar",
  "currentItemBackgroundColor": "transparent",
  "currentItemEndArrowhead": "arrow",
  "currentItemFillStyle": "hachure",
  "currentItemFontFamily": 1,
  "currentItemFontSize": 20,
  "currentItemLinearStrokeSharpness": "round",
  "currentItemOpacity": 100,
  "currentItemRoughness": 1,
  "currentItemStartArrowhead": null,
  "currentItemStrokeColor": "#000000",
  "currentItemStrokeSharpness": "sharp",
  "currentItemStrokeStyle": "solid",
  "currentItemStrokeWidth": 1,
  "currentItemTextAlign": "left",
  "cursorButton": "up",
  "draggingElement": null,
  "editingElement": null,
  "editingGroupId": null,
  "editingLinearElement": null,
  "elementLocked": false,
  "elementType": "selection",
  "errorMessage": null,
  "exportBackground": true,
  "exportEmbedScene": false,
  "exportPadding": undefined,
  "exportScale": 1,
  "exportWithDarkMode": false,
  "fileHandle": null,
  "gridSize": null,
  "isBindingEnabled": true,
  "isLibraryOpen": false,
  "isLoading": false,
  "isResizing": false,
  "isRotating": false,
  "lastPointerDownWith": "mouse",
  "metadata": "",
  "multiElement": null,
  "name": "name",
  "openMenu": null,
  "openPopup": null,
  "pasteDialog": Object {
    "data": null,
    "shown": false,
  },
  "previousSelectedElementIds": Object {},
  "resizingElement": null,
  "scrollX": 0,
  "scrollY": 0,
  "scrolledOutside": false,
  "selectedElementIds": Object {},
  "selectedGroupIds": Object {},
  "selectionElement": null,
  "shouldCacheIgnoreZoom": false,
  "showHelpDialog": false,
  "showStats": false,
  "startBoundElement": null,
  "suggestedBindings": Array [],
  "theme": "light",
  "toastMessage": null,
  "viewBackgroundColor": "#ffffff",
  "viewModeEnabled": false,
  "zenModeEnabled": false,
  "zoom": Object {
    "translation": Object {
      "x": 0,
      "y": 0,
    },
    "value": 1,
  },
}
`;

exports[`exportToSvg with exportEmbedScene 1`] = `"<!-- payload-type:application/vnd.excalidraw+json --><!-- payload-version:2 --><!-- payload-start -->eyJ2ZXJzaW9uIjoiMSIsImVuY29kaW5nIjoiYnN0cmluZyIsImNvbXByZXNzZWQiOnRydWUsImVuY29kZWQiOiJ4nO1Sy27CMFx1MDAxMLz3K1wi94pEklx1MDAwNkK50VJVlar2wFx1MDAwMalVXHUwMDBmJt4kXHUwMDE2xlx1MDAwZbZ5XHUwMDE18e+1XHUwMDFkXHUwMDFh01xiqT9cdTAwMTBcdTAwMWYs7czujnc9x5sgQPpQXHUwMDAxXHUwMDFhXHUwMDA3XGL2XHUwMDE5ZpRIvEM9i29BKiq4oWJcdTAwMTcrsZGZyyy1rsb9Plx1MDAxM6agXHUwMDE0Stf5wGBcdTAwMDVcXCuT8WniIDi6+0KBULxcdTAwMTKcuHRHeIkoXHUwMDBl2+ib4E4uikajMEzidNhkUDU1alx1MDAxYYihc8xcdTAwMTR4xkJoO5fr90m+pJzE62yVvk4nz1x1MDAxZl42p4zN9IHVw+Cs3EjwrNJSLGFOiS6telx1MDAwYm/qlDDL8lVSbIqSg1J/akSFM6pcdTAwMGZcdTAwMTZcdTAwMGL9gJhcdTAwMTeuh0f2JkqipIltRXyXtrRcdTAwMWZcdTAwMDVcdTAwMTPSat+G7nj1XHUwMDA1zpaFeVx1MDAwMic+J1x1MDAxYWC8yH3O7jxRfKFTXHUwMDAyLUrdXHUwMDAyXHUwMDE1uMVGYVx1MDAxMlxyXHUwMDA36X04alx1MDAxOCtRvVx1MDAxMPfFX+3FlFhW51x1MDAwNSBlg4vn2Zc91f74LXfcqXfdKcBcdTAwMTitXHUwMDE0dE7pnPKPUyRkut5S55XOK2evmNs1QriqZlx1MDAxYWu7lto5RoySXHUwMDE5/bZcYt8wVjdDW1xuu4cr0+buoFx1MDAxYtvz9Fx1MDAwM/n+1MkifQ==<!-- payload-end -->"`;
